passes/sat/sim.cc

```diff
@@ -2397,6 +2397,170 @@ struct VCDWriter : public OutputWriter
+		std::vector<uint32_t> lastValues;
+		std::vector<uint32_t> toggleCounts;
+		std::vector<uint32_t> highTimes;
+	};
+
+	typedef std::map<int, SignalActivityData> SignalActivityDataMap;
+
+	void write(std::map<int, bool> &use_signal) override
+	{
+		// Init map
+		SignalActivityDataMap dataMap;
+		// For each event (new time when a value changed)
+		for (auto &d : worker->output_data) {
+			// For each signal/values in that time slice
+			for (auto &data : d.second) {
+				int sig = data.first;
+				if (!use_signal.at(sig))
+					continue;
+			  // Create an entry in the map with all zeros for all bits of the signal
+				SignalActivityDataMap::iterator itr = dataMap.find(sig);
+				if (itr == dataMap.end()) {
+					Const value = data.second;
+					std::vector<uint32_t> vals(GetSize(value), 0);
+					SignalActivityData data;
+					data.highTimes = vals;
+					data.lastValues = vals;
+					data.toggleCounts = vals;
+					dataMap.emplace(sig, data);
+				}
+			}
+		}
+		// Max simulation time
+		int max_time = 0;
+		// clock pin id (highest toggling signal)
+		int clk = 0;
+		int highest_toggle = 0;
+		// Used to compute time intervals
+		int prev_time = 0;
+		// For each event (new time when a value changed)
+		for (auto &d : worker->output_data) {
+			int time = d.first;
+			// For each signal/values in that time slice
+			for (auto &data : d.second) {
+				int sig = data.first;
+				if (!use_signal.at(sig))
+					continue;
+				Const value = data.second;
+				SignalActivityDataMap::iterator itr = dataMap.find(sig);
+				std::vector<uint32_t> &lastVals = (*itr).second.lastValues;
+				std::vector<uint32_t> &toggleCounts = (*itr).second.toggleCounts;
+				std::vector<uint32_t> &highTimes = (*itr).second.highTimes;
+				for (int i = GetSize(value) - 1; i >= 0; i--) {
+					int val = '-';
+					switch (value[i]) {
+					case State::S0:
+						val = '0';
+						break;
+					case State::S1:
+						val = '1';
+						break;
+					case State::Sx:
+						val = 'x';
+						break;
+					default:
+						val = 'z';
+					}
+					if (val != lastVals[i]) {
+						toggleCounts[i]++;
+						if (toggleCounts[i] > highest_toggle) {
+							highest_toggle = toggleCounts[i];
+							clk = sig;
+						}
+						lastVals[i] = val;
+					}
+					if (lastVals[i] == '1') {
+						highTimes[i] += time - prev_time;
+					}
+				}
+			}
+			prev_time = time;
+			max_time = time;
+		}
+
+		// Retrieve VCD timescale
+		std::string timescale = worker->timescale;
+		double real_timescale = 1e-12; // ps
+		if (timescale == "ns")
+			real_timescale = 1e-9;
+		if (timescale == "fs")
+			real_timescale = 1e-15;
+
+		bool debug = false;
+
+		// Compute clock period, find the highest toggling signal and compute its average period
+		SignalActivityDataMap::iterator itr = dataMap.find(clk);
+		std::vector<uint32_t> &clktoggleCounts = (*itr).second.toggleCounts;
+		double clk_period = real_timescale * (double)max_time / (clktoggleCounts[0] / 2);
+		if (debug) {
```

Comment:
Use `log_debug` and set `options.log_level` in `preqorsor.json` instead?

Reply:
This is local debug that can go away in a couple of days after running all regressions.

Reply:
Ok. Put a TODO to remove anything that should be removed